incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,182 @@ type HierarchicalNamespaceReconciler struct {
+		log.Info("The self-serve subnamespace does not exist", "namespace", nm)
+		r.syncMissing(log, inst, ns)
+	case nsInst.Annotations[api.AnnotationOwner] != pnm:
+		// If the namespace is not a self-serve namespace or it's a self-serve namespace of another
+		// namespace. Report the conflict.
+		log.Info("The owner annotation of the subnamespace doesn't match the owner", "annotation", nsInst.Annotations[api.AnnotationOwner])
+		inst.Status.State = api.Conflict
+	default:
+		log.Info("The subnamespace has the correct owner annotation", "annotation", nsInst.Annotations[api.AnnotationOwner])
+		r.syncExisting(log, inst, ns, hcInst, nsInst)
+	}
+}
+
+func (r *HierarchicalNamespaceReconciler) syncMissing(log logr.Logger, inst *api.HierarchicalNamespace, ns *forest.Namespace) {
+	pnm := inst.Namespace
+	nm := inst.Name
+
+	// Set the HNS state to "Missing" because the subnamespace doesn't exist.
+	inst.Status.State = api.Missing
+
+	// Set the "Owner" in the forest of the hierarchical namespace to the current namespace.
+	log.Info("Setting the subnamespace's owner in the forest", "owner", pnm, "namespace", nm)
 	// TODO rename RequiredChildOf to Owner in the forest. See issue:
 	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/469
-	r.forest.Get(nm).RequiredChildOf = pnm
+	ns.RequiredChildOf = pnm
 
-	// Enqueue the in-momery hierarchyConfig instance of the hierarchical namespace.
-	// The hierarchyConfig reconciler will create the namespace and hierarchyConfig
-	// instances on apiserver accordingly.
+	// Enqueue the not-yet existent self-serve subnamespace. The HierarchyConfig reconciler will
+	// create the namespace and the HierarchyConfig instances on apiserver accordingly.
 	reason := "new/updated hierarchical namespace"
 	r.hcr.enqueueAffected(log, reason, nm)
+}
+
+// syncExisting syncs the existing subnamespace with its owner namespace. It updates the HNS state
+// to "Ok" or "Conflict" according to the hierarchy.
+func (r *HierarchicalNamespaceReconciler) syncExisting(log logr.Logger, inst *api.HierarchicalNamespace, ns *forest.Namespace, hcInst *api.HierarchyConfiguration, nsInst *corev1.Namespace) {
+	pnm := inst.Namespace
+	nm := inst.Name
+
+	switch hcInst.Spec.Parent {
+	case "":
+		log.Info("Parent is not set", "namespace", nm)
+		// This case is rare. It means the namespace is created with the right annotation but
+		// no HC. This could be a transient state before HC reconciler finishes creating the HC
+		// or a human manually created the namespace with the right annotation but no HC.
+		// Both cases meant to create this namespace as HNS.
+		log.Info("Setting the subnamespace's owner in the forest", "owner", pnm, "namespace", nm)
+		ns.RequiredChildOf = pnm
 
-	// TODO sync with forest to report conflicts in hns.Status.State. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/487
+		reason := "new/updated hierarchical namespace"
+		r.hcr.enqueueAffected(log, reason, nm)
```

Comment:
Shouldn't this just be "updated" (not "new")?

Also feel free to combine into one line if you like: `r.hcr.enqueueAffected(log, "updated hierarchical namespace", nm)`